storagenode/gracefulexit/worker.go

```diff
@@ -123,6 +125,41 @@ func (worker *Worker) Run(ctx context.Context, done func()) (err error) {
 				continue
 			}
 
+			if !bytes.Equal(originalHash.Hash, pieceHash.Hash) {
+				worker.log.Error("piece hash from new storagenode does not match", zap.Stringer("storagenode ID", addrLimit.Limit.StorageNodeId), zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID))
+				worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+				continue
+			}
+			if !bytes.Equal(pieceHash.PieceId.Bytes(), addrLimit.Limit.PieceId.Bytes()) {
+				worker.log.Error("piece id from new storagenode does not match order limit", zap.Stringer("storagenode ID", addrLimit.Limit.StorageNodeId), zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID))
+				worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+				continue
+			}
+
+			// ecclient.PutPiece already dials the node, but we need to access the connection directly to check the signature
+			nodeConn, err := worker.dialer.DialNode(ctx, &pb.Node{
```

Comment:
Do not dial again, we need to avoid adding load to the storagenodes, unless it's really necessary.

Reply:
Fixed